com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/AcfKeywordHelper.java

```diff
@@ -0,0 +1,293 @@
+   *
+   * @param context
+   *          Some grammar element
+   * @return Keyword helper
+   */
+  public static AcfKeywordHelper getHelper(final EObject context) {
+    for (Adapter candidate : context.eResource().getResourceSet().eAdapters()) {
+      if (candidate instanceof AcfKeywordHelper) {
+        return (AcfKeywordHelper) candidate;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns a rule name for the given keyword value.
+   *
+   * @param keywordValue
+   *          Keyword value
+   * @return Rule name
+   */
+  public String getRuleName(final String keywordValue) {
+    return keywordValueToToken.get(createKey(keywordValue));
+  }
+
+  /**
+   * Returns keyword value for the given rule name.
+   *
+   * @param ruleName
+   *          Rule name
+   * @return Keyword value
+   */
+  public String getKeywordValue(final String ruleName) {
+    return keywordValueToToken.inverse().get(ruleName).toString();
+  }
+
+  /**
+   * Checks whether the rule with the given name is a keyword rule.
+   *
+   * @param ruleName
+   *          Name of a rule
+   * @return {@code true} if the rule is a keyword rule, {@code false} otherwise
+   */
+  public boolean isKeywordRule(final String ruleName) {
+    return keywordValueToToken.containsValue(ruleName);
+  }
+
+  /**
+   * Returns all keywords.
+   *
+   * @return Set of keywords as strings
+   */
+  public Set<String> getAllKeywords() {
+    Set<String> result = new TreeSet<>(KeywordHelper.keywordComparator);
+    for (CharSequence cs : keywordValueToToken.keySet()) {
+      result.add(cs.toString());
+    }
+
+    return Collections.unmodifiableSet(result);
+  }
+
+  /**
+   * Initializes the keyword map.
+   * Searches for keywords that belong to grammar rules reachable from the first rule of the given grammar.
+   *
+   * @param grammar
+   *          Root grammar
+   * @return Keyword map
+   */
+  private BiMap<CharSequence, String> createKeywordMap(final Grammar grammar) {
+    List<ParserRule> parserRules = ParserSemanticPredicatesUtil.allReachableParserRules(grammar);
+    List<EnumRule> enumRules = ParserSemanticPredicatesUtil.allReachableEnumRules(grammar);
+    Iterator<EObject> iter = Iterators.concat(EcoreUtil.<EObject> getAllContents(parserRules), EcoreUtil.<EObject> getAllContents(enumRules));
+    Iterator<Keyword> filtered = Iterators.filter(iter, Keyword.class);
+    Iterator<String> transformed = Iterators.transform(filtered, new Function<Keyword, String>() {
+      @Override
+      public String apply(final Keyword from) {
+        return from.getValue();
+      }
+    });
+    TreeSet<String> treeSet = Sets.newTreeSet(new Comparator<String>() {
+      @Override
+      public int compare(final String o1, final String o2) {
+        if (o1.length() == o2.length()) {
+          return o1.compareTo(o2);
+        }
+        return Integer.valueOf(o1.length()).compareTo(Integer.valueOf(o2.length()));

```

Comment:
There is no need to create this Integer objects, easier to read and faster is
Integer.compare(o1.length(), o2.length())

Reply:
ok.